packages/lab/src/list/useList.ts

```diff
@@ -1,569 +1,244 @@
-    },
-    " ": (event) => {
-      event.preventDefault();
-      handleSelect(event, highlightedIndex, getItemAtIndex(highlightedIndex));
-    },
-    Tab: (event) => {
-      if (tabToSelect) {
-        handleSelect(event, highlightedIndex, getItemAtIndex(highlightedIndex));
-      } else {
-        setHighlightedIndex(undefined);
-      }
-    },
-  };
-
-  const handleKeyDown: keyHandler = (event) => {
-    if (isFocusVisibleRef.current) {
-      setFocusVisible(true);
-    }
-
-    const handler: keyHandler = keyDownHandlers[event.key];
-
-    if (handler) {
-      handler(event);
-    }
-
-    if (onKeyDown) {
-      onKeyDown(event);
-    }
-  };
-
-  const handleFocus = (event: FocusEvent<HTMLDivElement>) => {
-    handleFocusVisible(event);
-    if (isFocusVisibleRef.current) {
-      setFocusVisible(true);
-    }
-
-    // Work out the index to highlight
-    if (highlightedIndex === undefined) {
-      const firstSelectedItem = isMultiSelect
-        ? (selectedItem as Item[])[0]
-        : selectedItem;
-
-      setHighlightedIndex(
-        Math.max(
-          restoreLastFocus ? lastFocusedIndex : getItemIndex(firstSelectedItem),
-          0
-        )
-      );
-    }
-
-    if (onFocus) {
-      onFocus(event);
-    }
-  };
-
-  const handleBlur = (event: FocusEvent<HTMLDivElement>) => {
-    setHighlightedIndex(undefined);
-    handleBlurVisible();
-    if (!isFocusVisibleRef.current) {
-      setFocusVisible(false);
-    }
-
-    if (onBlur) {
-      onBlur(event);
-    }
-  };
-
-  const handleMouseDown = (event: MouseEvent<HTMLDivElement>) => {
-    if (disableMouseDown) {
-      event.preventDefault();
-    } else if (onMouseDown) {
-      onMouseDown(event);
-    }
-  };
-
-  const handleMouseLeave = (event: MouseEvent<HTMLDivElement>) => {
-    if (focusVisible) {
-      // Get the root node of the component if we have access to it otherwise default to current document
-      const rootNode = (
-        rootRef.current || ownerDocument(event.currentTarget)
-      ).getRootNode();
-
-      const listNode = (rootNode as Document).getElementById(id);
-
-      // Safety check as `mouseleave` could have been accidentally triggered by an opening tooltip
-      // when you use keyboard to navigate, hence the focusVisible check earlier
-      if (listNode?.contains(event.target as Node)) {
-        setHighlightedIndex(undefined);
-      }
-    } else {
-      setHighlightedIndex(undefined);
-    }
-
-    if (onMouseLeave) {
-      onMouseLeave(event);
-    }
-  };
-
-  const eventHandlers = {
-    onFocus: handleFocus,
-    onBlur: handleBlur,
+  const getActiveDescendant = () =>
+    highlightedIndex === undefined ||
+    highlightedIndex === -1 ||
+    disableAriaActiveDescendant
+      ? undefined
+      : dataHook.data[highlightedIndex]?.id;
+
+  // We need this on reEntry for navigation hook to handle focus
+  lastSelection.current = selectionHook.selected;
```

Comment:
This is a tricky one. I don't like the pattern I'm using and would welcome a better approach. I think it should still work with repeated renders a la React 18, though.

The problem I'm trying to solve is a circular dependency between the keyboard navigation hook and the selection hook. KeyboardNav needs to know the current selection, because we sometimes need to  begin navigation from the selected item.

But, selection only gets set after the keyboard nav hook is called and the selection hook needs the highlighted index which is returned from the keyboard nav hook. It needs this so that if Enter key is pressed, it selects the highlighted item.

So, we store the selection in a ref, initialised to the default value before the keyboard nav hook is called. It works because navigation always operates from the previous selection and navigation never changes selection. The selection hook must follow navigation, because selection must operate on the most up to fate highlighted index.

after both hooks have run, we set the selected ref to point to the current selection, ready for next render, whenever that happens.